### PR TITLE
Remove `useId` prefix

### PR DESCRIPTION
**Pull request by lensbart**
# Allow `useId()` (i.e., without argument)

No ticket, but previously mentioned [in this Slack comment](https://datacamp.slack.com/archives/CHT1GJV0E/p1659022911822859)

`useId` had a mandatory argument that is used as a prefix for the generated id. In my view, this is redundant:
* The main benefit of using something like `useId` is that you can forget about the actual id being generated. Having to provide a prefix goes against this idea.
* `useId` is typically assigned to a constant, so the developer can already choose a nice variable name to clarify purpose.
* React 18 provides a hook `useId` that doesn’t take an argument, so it would be nice to be API compatible so that we can more easily swap the Waffles `useId` with the React one, once we upgrade our `react` dependency.
* If we want to have guaranteed no naming collisions, we should either use more random bytes or follow React’s approach using `identifierPrefix`. But I don’t think this was the reason for enforcing a prefix in the first place.

## Changes

* Makes the argument to `useId` optional
* The previous implementation used `useState` for something that was not actually state. It also came with one redundant rerender. I’ve simplified the code to use `useMemo`.
* Changes all examples so that they don’t suggest providing a prefix anymore (with the exception of one example for `useId`, precisely to exemplify the use of a prefix). Absent any tradeoff, I think we should encourage simplicity. I’ve done this in a separate commit, in case you disagree.

### 🔧 Minor Changes

- Argument to `useId` is now optional

### ❓ Misc Changes

- Argument to `useId` is not encouraged anymore via examples in docs

## Question

I would prefer to drop the example with an argument altogether, because
* I don’t see any use case for this
* This way, the migration path to React 18’s `useId` will be smoother.

I think we should deprecate using `useId` with an argument, but I held off for now, pending your feedback.

Thank you!

### CODE REVIEW

File: doc-site/examples/hooks/use-id-basic-usage.tsx [hunk 1]
```diff
@@ -10,11 +10,20 @@ const errorStyle = css`
 `;
 
 function Example() {
+  const descriptionId = useId();
```
Comment (lensbart) [inline doc-site/examples/hooks/use-id-basic-usage.tsx:13]: Ideally I’d just edit the existing example to `const errorId = useId();`, but if there is a use case to providing `'input-error'` that I’m not seeing, I guess providing both examples is fine.

**Comment by blurbyte** [comment]
I agree the prefix should be optional. Personally I think it's still nice to have this prefix, and not sure if its usage should be discouraged: so instead of 
```
const inputId = useId();
// Somewhere in the component
<input id={`my-input-${inputId}`} />
```
we can write
```
const inputId = useId('my-input`);
// Somewhere in the component
<input id={inputId} />
```
Of course it's such a small detail I guess we can live without it.

On the other hand without prefix it will be more in line with React 18 `useId`. So I am kind of split on it.
@ixTec Any opinion on it?

**Comment by ixTec** [comment]
I'm leaning towards no prefix to align with react 18, and we can always use something like ``const id = `prefix-${useId()}`;`` if we still want a prefix internally etc.

However, changing it now would be a fairly large breaking change to the use of it and so I think we reserve that for our next **major** release Waffles v2.0.0, and just make it optional for now in the meantime

**Comment by lensbart** [comment]
> so instead of
> 
> ```
> const id = useId();
> // Somewhere in the component
> <input id={`my-input-${id}`} />
> ```
> 
> we can write
> 
> ```
> const inputId = useId('my-input`);
> // Somewhere in the component
> <input id={inputId} />
> ```

What’s the use case for having `my-input-${id}`? (Apart from the example in the React docs, because that one presupposes one `useId` call for multiple elements within a component.)

**Comment by blurbyte** [comment]
There are a few similar use cases in our code base we can follow this pattern (e.g. `FormField` component).
So what I'd like to suggest right now is to **drop prefix altogether** and make it similar to React 18 `useId`. Technically it's breaking change but in reality `useId` is used mostly as internal utility, and Workspace team will be aware of this change as well. In docs I would show how the prefix could be added via component props.
@lensbart Is it something you would like to work on? It would require to update & test at least a few Waffles components.

**Comment by lensbart** [comment]
> Is it something you would like to work on?

This week I’m quite busy (taking Friday off), I might look into it next week if it’s not already done by then.

**Comment by blurbyte** [comment]
Ok, I can do it then, not a LOT of work tbh.

**Comment by lensbart** [comment]
Also, the return type would ideally be `string` instead of `string | undefined`

**Comment by blurbyte** [comment]
@ixTec @lensbart

Made some improvements to `useId` as we discussed:

- Removed prefix arg, so it's in line with React 18 built-in `useId`.
- If React 18 is available built-in `useId` is used.
- It works on the server side too (aka no warning about IDs mismatch should appear). At first on the server and in first rendering pass it's _null_, and after that it's patched to actual ID, so performance is not ideal but should be negligible). Of course React 18 useId don't have this double rendering downside.
- Left prefixes in most Waffles components, since it makes testing a bit easier and deterministic when you have one component using multiple IDs. But be free to NOT use it if you feel you don't need it - we don't either discourage / encourage doing so.

### 🚨 Breaking Changes

- `useId` no longer accepts _prefix_ argument
- removed optional `id` from `FormField` since it's no longer needed

**Review by ixTec — CHANGES_REQUESTED** [review]

File: src/form-field/form-field.tsx [hunk 2]
```diff
@@ -40,9 +37,9 @@ function FormField({
   error,
   children,
 }: FormFieldProps) {
-  const generatedId = useId('form-field');
-  const fieldId = id || generatedId;
```
[the remark this answers is not in the document]
Comment (blurbyte) [reply src/form-field/form-field.tsx:44]: This optional `id` was here only if someone would be bothered with this 'IDs mismatch between server and client' warning. But now hook handles transition between server and client so it should be fine.

ID is injected into child element like this (line 44 of _form-field.tsx_):
```
const element = cloneElement(children, {
    id: fieldId,
    ...
 });
```
If you go to examples everything works fine (when you click label input got focused).

[the remark this answers is not in the document]
Comment (ixTec) [reply src/form-field/form-field.tsx:44]: Oops, I clearly need more coffee because I missed that line within our `cloneElement` - disregard my comment :D

**Review by ixTec — APPROVED** [review]

**Pull request merged**